torch/optim/adagrad.py

```diff
@@ -56,14 +59,11 @@ def step(self, closure=None):
 
                 state['step'] += 1
 
-                if group['weight_decay'] != 0:
-                    if p.grad.data.is_sparse:
-                        raise RuntimeError("weight_decay option is not compatible with sparse gradients")
-                    grad = grad.add(group['weight_decay'], p.data)
-
                 clr = group['lr'] / (1 + (state['step'] - 1) * group['lr_decay'])
 
                 if grad.is_sparse:
+                    if group['weight_decay'] != 0:
+                        raise RuntimeError("weight_decay option is not compatible with sparse gradients")
```

Comment:
The new weight decay update is added directly to p.data, so I think sparse gradients are no longer a problem for weight_decay.